### ISSUE TITLE
trouble with titles in epubsplit.py

### ISSUE THREAD

**Issue by cognitivetech** [JimmXinu/EpubSplit#11]
really loving the script, and it works well in most cases.. but having an issue with a title that I've been fighting with for days...

I isolated from your code a script to check metadata of a given title... which is where I realized that this book has titles that can be found by your script.. but for some reason aren't printed as lines.

here is the section I added to main() of `epubsplit.py`

```python
    # Print TOC from original get_toc_map
    print("\nTOC Map:")
    original_toc_map = epubO.get_toc_map()
    for href, toc_items in original_toc_map.items():
        for text, anchor in toc_items:
            print(f"  - {text} (href: {href}, anchor: {anchor})")
```

when I use this method, I get the title... but just running the unmodified script without options gives me no titles. (and the problem with mine is it gets no line numbers, for eventual extraction)

```bash
$ python emeta2.py ~/jamstack/summarization/summaries/source_documents/shortlist/They\ Say\ I\ Say\ \(Gerald\ Graff\)\ \(Z-Library\).epub 

Line Number: 0
        guide: ('text', 'Beginning')
        id: npp-794c427b36dac68b15a0
        href: EPUB/content/0.1.0-cover-tsis.xhtml

Line Number: 1
        id: npp-ee00549451c2f9991260
        href: EPUB/content/0.2-publishersnotice.xhtml

Line Number: 2
        id: npp-2aeea85cccb9fb635f3b
        href: EPUB/content/0.2.0-praise-tsis.xhtml

Line Number: 3
        id: npp-6e77de5ddd05c33fcfd8
        href: EPUB/content/0.3.0-title-tsis.xhtml

Line Number: 4
        id: npp-288a5c30919f6be35d5f
        href: EPUB/content/0.4.0-copyright-tsis.xhtml

Line Number: 5
        id: npp-52ab84dbc2d276389212
        href: EPUB/content/0.5.0-dedication-tsis.xhtml

Line Number: 6
        id: npp-dca9e38a10e4e2a75a84
        href: EPUB/content/0.6.0-contents-tsis.xhtml

Line Number: 7
        id: npp-690cb30cf7a6a880da56
        href: EPUB/content/0.7.0-preface-edition-tsis.xhtml

Line Number: 8
        id: npp-3b93a8ae73c7cd1e4f9c
        href: EPUB/content/0.8.0-preface-tsis.xhtml

Line Number: 9
        id: npp-5416d0e7dc616fc1700e
        href: EPUB/content/0.9.0-introduction-tsis.xhtml

Line Number: 10
        id: npp-3a5247923fa827aead13
        href: EPUB/content/1.0.0-part01.xhtml

Line Number: 11
        id: npp-8a0aa6284f30d5ecc958
        href: EPUB/content/1.1.0-chapter01.xhtml

Line Number: 12
        id: npp-156bdeb8013c205f91ab
        href: EPUB/content/1.2.0-chapter02.xhtml

Line Number: 13
        id: npp-c784b3f74bee000b1ae5
        href: EPUB/content/1.3.0-chapter03.xhtml

Line Number: 14
        id: npp-ff12b52c84bb1fb532c9
        href: EPUB/content/2.0.0-part02.xhtml

Line Number: 15
        id: npp-e01ad8a4fcb771af5844
        href: EPUB/content/2.1.0-chapter04.xhtml

Line Number: 16
        id: npp-9a995f0313318a9b9af2
        href: EPUB/content/2.2.0-chapter05.xhtml

Line Number: 17
        id: npp-390681278e71a4423ed9
        href: EPUB/content/2.3.0-chapter06.xhtml

Line Number: 18
        id: npp-f5c4e1a9dd901a287b28
        href: EPUB/content/2.4.0-chapter07.xhtml

Line Number: 19
        id: npp-68c81eb6520d60eefaa5
        href: EPUB/content/3.0.0-part03.xhtml

Line Number: 20
        id: npp-ddbc1555801edba7bf3d
        href: EPUB/content/3.1.0-chapter08.xhtml

Line Number: 21
        id: npp-b9dd52ac4529d15138ba
        href: EPUB/content/3.2.0-chapter09.xhtml

Line Number: 22
        id: npp-f78033bbb0012b711c93
        href: EPUB/content/3.3.0-chapter10.xhtml

Line Number: 23
        id: npp-e24bf11a2e1d49b077d4
        href: EPUB/content/3.4.0-chapter11.xhtml

Line Number: 24
        id: npp-0daa12a661b5830c3527
        href: EPUB/content/4.0.0-part04.xhtml

Line Number: 25
        id: npp-7882e2a49e3a34d30075
        href: EPUB/content/4.1.0-chapter12.xhtml

Line Number: 26
        id: npp-4a5d72e8a3d4345c81a3
        href: EPUB/content/4.2.0-chapter13.xhtml

Line Number: 27
        id: npp-55773e835f66c9f12a0d
        href: EPUB/content/4.3.0-chapter14.xhtml

Line Number: 28
        id: npp-ddc2552e8a089a751e94
        href: EPUB/content/4.3.1-chapter14.xhtml

Line Number: 29
        id: npp-ba3840a954dbcd5cc043
        href: EPUB/content/4.4.0-chapter15.xhtml

Line Number: 30
        id: npp-7015bbf5802d65fa5af0
        href: EPUB/content/4.5.0-chapter16.xhtml

Line Number: 31
        id: npp-51ef09d5ca41ef4c2e03
        href: EPUB/content/4.6.0-chapter17.xhtml

Line Number: 32
        id: npp-8c901cc6e92f02b90a0d
        href: EPUB/content/4.7.0-chapter18.xhtml

Line Number: 33
        id: npp-f2b9be171d933026a88a
        href: EPUB/content/5.0.0-readings-tsis.xhtml

Line Number: 34
        id: npp-f79b75a8d1e698015ddc
        href: EPUB/content/5.1.0-reading-tsis.xhtml

Line Number: 35
        id: npp-4d4eb48143dc81fa4507
        href: EPUB/content/5.2.0-reading-tsis.xhtml

Line Number: 36
        id: npp-3dd45822f6a716727e1b
        href: EPUB/content/5.3.0-reading-tsis.xhtml

Line Number: 37
        id: npp-85f1d809202646b9c472
        href: EPUB/content/5.4.0-reading-tsis.xhtml

Line Number: 38
        id: npp-0a168f7c9d800fc2be83
        href: EPUB/content/5.5.0-reading-tsis.xhtml

Line Number: 39
        id: npp-b36b1cd1de857bd42290
        href: EPUB/content/6.0.0-credits-tsis.xhtml

Line Number: 40
        id: npp-33ead1344f3d9fd24d8f
        href: EPUB/content/7.0.0-acknowledgments-tsis.xhtml

Line Number: 41
        id: npp-33420409bea0f1e1dd07
        href: EPUB/content/8.0.0-index-of-templates-tsis.xhtml

Line Number: 42
        id: npp-95280ae93efde337a8f6
        href: EPUB/content/9.0.0-about-the-authors-tsis.xhtml

TOC Map:
  - COVER (href: EPUB/EPUB/content/0.1.0-cover-tsis.xhtml, anchor: None)
  - TITLE PAGE (href: EPUB/EPUB/content/0.3.0-title-tsis.xhtml, anchor: None)
  - CONTENTS (href: EPUB/EPUB/content/0.6.0-contents-tsis.xhtml, anchor: None)
  - PREFACE TO THE FIFTH EDITION (href: EPUB/EPUB/content/0.7.0-preface-edition-tsis.xhtml, anchor: None)
  - PREFACE: Demystifying Academic Conversation (href: EPUB/EPUB/content/0.8.0-preface-tsis.xhtml, anchor: None)
  - INTRODUCTION: Entering the Conversation (href: EPUB/EPUB/content/0.9.0-introduction-tsis.xhtml, anchor: None)
  - PART 1 “THEY SAY” (href: EPUB/EPUB/content/1.0.0-part01.xhtml, anchor: None)
  - ONE “THEY SAY”: Starting with What Others Are Saying (href: EPUB/EPUB/content/1.1.0-chapter01.xhtml, anchor: None)
  - TWO “HER POINT IS”: The Art of Summarizing (href: EPUB/EPUB/content/1.2.0-chapter02.xhtml, anchor: None)
  - THREE “AS HE HIMSELF PUTS IT”: The Art of Quoting (href: EPUB/EPUB/content/1.3.0-chapter03.xhtml, anchor: None)
  - PART 2 “I SAY” (href: EPUB/EPUB/content/2.0.0-part02.xhtml, anchor: None)
  - FOUR “YES / NO / OK, BUT”: Three Ways to Respond (href: EPUB/EPUB/content/2.1.0-chapter04.xhtml, anchor: None)
  - FIVE “AND YET”: Distinguishing What You Say from What They Say (href: EPUB/EPUB/content/2.2.0-chapter05.xhtml, anchor: None)
  - SIX “SKEPTICS MAY OBJECT”: Planting a Naysayer in Your Text (href: EPUB/EPUB/content/2.3.0-chapter06.xhtml, anchor: None)
  - SEVEN “SO WHAT? WHO CARES”: Saying Why It Matters (href: EPUB/EPUB/content/2.4.0-chapter07.xhtml, anchor: None)
  - PART 3 TYING IT ALL TOGETHER (href: EPUB/EPUB/content/3.0.0-part03.xhtml, anchor: None)
  - EIGHT “AS A RESULT”: Connecting the Parts (href: EPUB/EPUB/content/3.1.0-chapter08.xhtml, anchor: None)
  - NINE “YOU MEAN I CAN JUST SAY IT THAT WAY?”: Academic Writing Doesn’t Mean Setting Aside Your Own Voice (href: EPUB/EPUB/content/3.2.0-chapter09.xhtml, anchor: None)
  - TEN “BUT DON’T GET ME WRONG”: The Art of Metacommentary (href: EPUB/EPUB/content/3.3.0-chapter10.xhtml, anchor: None)
  - ELEVEN “WHAT I REALLY WANT TO SAY IS”: Revising Substantially (href: EPUB/EPUB/content/3.4.0-chapter11.xhtml, anchor: None)
  - PART 4 IN SPECIFIC ACADEMIC CONTEXTS (href: EPUB/EPUB/content/4.0.0-part04.xhtml, anchor: None)
  - TWELVE “I TAKE YOUR POINT”: Entering Class Discussions (href: EPUB/EPUB/content/4.1.0-chapter12.xhtml, anchor: None)
  - THIRTEEN DON’T MAKE THEM SCROLL UP: Entering Online Conversations (href: EPUB/EPUB/content/4.2.0-chapter13.xhtml, anchor: None)
  - FOURTEEN WHAT’S MOTIVATING THIS WRITER?: Reading for the Conversation (href: EPUB/EPUB/content/4.3.0-chapter14.xhtml, anchor: None)
  - David Zinczenko, Don’t Blame the Eater (href: EPUB/EPUB/content/4.3.1-chapter14.xhtml, anchor: None)
  - FIFTEEN “BUT AS SEVERAL SOURCES SUGGEST”: Research as Conversation (href: EPUB/EPUB/content/4.4.0-chapter15.xhtml, anchor: None)
  - SIXTEEN “ON CLOSER EXAMINATION”: Entering Conversations about Literature (href: EPUB/EPUB/content/4.5.0-chapter16.xhtml, anchor: None)
  - SEVENTEEN “THE DATA SUGGEST”: Writing in the Sciences (href: EPUB/EPUB/content/4.6.0-chapter17.xhtml, anchor: None)
  - EIGHTEEN “ANALYZE THIS”: Writing in the Social Sciences (href: EPUB/EPUB/content/4.7.0-chapter18.xhtml, anchor: None)
  - READINGS (href: EPUB/EPUB/content/5.0.0-readings-tsis.xhtml, anchor: None)
  - Gerald Graff, Hidden Intellectualism (href: EPUB/EPUB/content/5.1.0-reading-tsis.xhtml, anchor: None)
  - Michelle Alexander, The New Jim Crow: Mass Incarceration in the Age of Colorblindness (href: EPUB/EPUB/content/5.2.0-reading-tsis.xhtml, anchor: None)
  - Kelly Coryell, All Words Matter: The Manipulation behind “All Lives Matter” (href: EPUB/EPUB/content/5.3.0-reading-tsis.xhtml, anchor: None)
  - Michael Littman, “Rise of the Machines” Is Not a Likely Future (href: EPUB/EPUB/content/5.4.0-reading-tsis.xhtml, anchor: None)
  - Gavin Reid, The Electoral College Embodies American Ideals (href: EPUB/EPUB/content/5.5.0-reading-tsis.xhtml, anchor: None)
  - CREDITS (href: EPUB/EPUB/content/6.0.0-credits-tsis.xhtml, anchor: None)
  - ACKNOWLEDGMENTS (href: EPUB/EPUB/content/7.0.0-acknowledgments-tsis.xhtml, anchor: None)
  - INDEX OF TEMPLATES (href: EPUB/EPUB/content/8.0.0-index-of-templates-tsis.xhtml, anchor: None)
  - ABOUT THE AUTHORS (href: EPUB/EPUB/content/9.0.0-about-the-authors-tsis.xhtml, anchor: None)
```

I'm sure this is probably another example of the epub going off spec...
do you have some special code you use to examine the sctructure of epub when encountering such issues?

I'm grateful for your time thinking about this, because I've been fighting with this issue for a couple days with no luck

**Comment by JimmXinu** [comment]
I use Calibre's Edit book feature to look inside the epubs directly.  Or if you don't use Calibre, you can always unzip it--an epub is just a zip file with some defined files and structure.

If you look closely, the 'Lines' each reference files like `EPUB/content/0.1.0-cover-tsis.xhtml` while the TOC map shows `EPUB/EPUB/content/0.1.0-cover-tsis.xhtml`.  Note the extra `EPUB/`.  So the lines from the file manifest don't match up with the TOC entries.

It's possible the TOC file (usually `toc.ncx`) is in an unusual, but valid, place, or that the TOC isn't valid.

If it's valid, I should fix EpubSplit.  I'd have to see the epub to say for sure.

**Comment by JimmXinu** [comment]
Thanks.  That epub is, in fact, valid.  It's also the first one I've seen where `toc.ncx` _isn't_ in the same relative path as `content.opf`, or `EPUB/package.opf` in this case.

I've pushed a change to explicitly use `toc.ncx` relative path instead that seems to work.

**Comment by cognitivetech** [comment]
brilliant! now when I run from the terminal, it shows split-lines with their appropriate titles...

unfortunately, when calling as a function I get same results as before

**Comment by JimmXinu** [comment]
I don't understand your issue at this point.

Before, TOC entries weren't mapping to manifest entries because of href mismatch due to the bad assumption `toc.ncx` and `content.opf` would always share a relative path.  So TOC entries weren't appearing in the 'lines' list, just the manifested files.

That's fixed in 4e5bf320951e02da7ad5ba456b75934340d2b4be.  `get_split_lines()` returns the raw data for lines output by the CLI, including TOC entries.
